markdown/developers/pipeline_structure.md

```diff
@@ -0,0 +1,74 @@
+
+* `docs`: Markdown files for documenting the pipeline
+
+* `lib`: The lib directory contains groovy utility functions. Currently, this includes the `Schema.groovy`, `Headers.groovy`, `Completion.groovy` and `Checks.groovy` files. These are called from within the nf-core pipeline to do groovy-related tasks, like formatting the header for the command line. Any larger groovy function you want to call in the code should therefore go in here.
+
+* `modules`: Contains local and nf-core modules
+
+## Continuous integration testing
+
+To assure that nf-core pipelines don't break after some change is made to the code, we use automated continuous integration (CI) testing. This is done via GitHub actions, which are defined in the `.github/workflows` directory. Parameters and file paths are set in the `conf/test.config` and `conf/test_full.config`. Please see also [here](https://nf-co.re/developers/adding_pipelines#add-some-test-data) for how to set-up the test workflow for your pipeline.
+
+## DSL2 and modules
+
+Nextflow DSL2 allows for a more modularized design of pipelines and the reuse of components. Currently, most nf-core pipelines are still entirely written in DSL1, but in the near future all pipelines will be written in DSL2. The nf-core team has developed a set of design patterns on how to best implement DSL2 pipelines, which should be used by all nf-core pipelines in order to assure standardization and the reuse of components. The following is meant to help understand certain design choices and how a nf-core DSL2 pipeline should be build.
+
+### modules
+
+Each pipeline has a `modules` directory which contains all the module code. A module here depicts a single process which involves - if possible - only a single tool/software. The `modules` directory is furthermore divided into `local`and `nf-core`sub-directories, which themselves each have a `process`/`software` and `subworkflow` directory. Modules contained in the `local` directory are specific to the pipeline, whereas `nf-core` modules are installed from the `nf-core/modules` repository. For instance, most pipelines that involve FastQ files will run the FastQC tool for quality control. The module needed for this can be easily reused from the `nf-core/modules` directory using the `nf-core/tools`package. The `process` directories contain modules which define single processes, which smaller workflows are contained in the `subworkflow` directories.
```

Comment:
```suggestion
Each pipeline has a `modules` directory which contains all the module code. A module here depicts a single process which involves - if possible - only a single tool/software. The `modules` directory is furthermore divided into `local`and `nf-core` sub-directories, which themselves each have a `process`/`software` and `subworkflow` directory. Modules contained in the `local` directory are specific to the pipeline, whereas `nf-core` modules are installed from the `nf-core/modules` repository. For instance, most pipelines that involve FastQ files will run the FastQC tool for quality control. The module needed for this can be easily reused from the `nf-core/modules` directory using the `nf-core/tools`package. The `process` directories contain modules which define single processes, which smaller workflows are contained in the `subworkflow` directories.
```

Reply:
Might be worth having a paragraph making it very clear that no files in the `nf-core` directory (from nf-core/modules) should ever be edited. Instead, use the `nf-core modules` cli commands to install, update and remove these files. Automatic code checks will fail if these differ from the central nf-core/modules repository at all.

Could also be good to go on to explain how versioning works. But I don't think that this is actually written yet, so maybe leave that until it exists 😆

Reply:
Yes good point!